### PR TITLE
feat: 🎸 json tree shaking

### PR DESCRIPTION
**Pull request by IWANABETHATGUY**
## Summary
1. Closed https://github.com/web-infra-dev/rspack/issues/4896
2. Closed https://github.com/web-infra-dev/rspack/issues/4942

## Require Documentation?

- [ ] No
- [ ] Yes, the corresponding rspack-website PR is \_\_

### CODE REVIEW

File: webpack-test/configCases/sharing/share-multiple-versions/webpack.config.js [hunk 1]
```diff
@@ -3,6 +3,9 @@ const { SharePlugin } = require("../../../../").sharing;
 
 /** @type {import("@rspack/core").Configuration} */
 module.exports = {
+	optimization: {
```
Comment (IWANABETHATGUY) [inline webpack-test/configCases/sharing/share-multiple-versions/webpack.config.js:6]: tracked https://github.com/web-infra-dev/rspack/issues/4935

File: webpack-test/cases/json/import-by-name/test.filter.js [hunk 2]
```diff
@@ -1,4 +1,4 @@
 
-module.exports = () => {return "https://github.com/web-infra-dev/rspack/issues/4323"}
```
Comment (IWANABETHATGUY) [inline webpack-test/cases/json/import-by-name/test.filter.js:2]: want to make sure json tree shaking generate valid result.

File: webpack-test/cases/json/import-by-name/warnings.js
Comment (IWANABETHATGUY) [inline webpack-test/cases/json/import-by-name/warnings.js:1]: Will recover after we support https://github.com/web-infra-dev/rspack/issues/4323

File: webpack-test/cases/json/import-by-name/test.filter.js (hunk 2 above)
Comment (ahabhgk) [reply webpack-test/cases/json/import-by-name/test.filter.js:2]: Try not to modify the webpack-test, so we can track the different behavior, if only for valid the result, you can create copy the test into `packages/rspack/tests`

File: crates/rspack_plugin_json/tests/fixtures/simple/snapshot/output.snap [hunk 3]
```diff
@@ -12,10 +12,7 @@ console.log(_json_json__WEBPACK_IMPORTED_MODULE_0__);
 }),
 "./json.json": (function (module, exports, __webpack_require__) {
 "use strict";
-module.exports = {
```
[the remark this answers is not in the document]
Comment (IWANABETHATGUY) [reply crates/rspack_plugin_json/tests/fixtures/simple/snapshot/output.snap:15]: content is still the same.

File: webpack-test/cases/json/import-by-name/test.filter.js (hunk 2 above)
Comment (IWANABETHATGUY) [reply webpack-test/cases/json/import-by-name/test.filter.js]: Done

File: crates/rspack_plugin_json/tests/fixtures/simple/snapshot/output.snap [hunk 4]
```diff
@@ -12,10 +12,7 @@ console.log(_json_json__WEBPACK_IMPORTED_MODULE_0__);
...
-  "hello": "world"
-}
-;}),
+module.exports = JSON.parse('{"hello":"world is better \'"}')}),
```
Comment (IWANABETHATGUY) [inline crates/rspack_plugin_json/tests/fixtures/simple/snapshot/output.snap:15]: Json optimization.https://github.com/webpack/webpack/blob/1f99ad6367f2b8a6ef17cce0e058f7a67fb7db18/lib/json/JsonGenerator.js#L186

**Review by ahabhgk — APPROVED** [review]

**Pull request merged**